contracts/MeanFinance/MeanFinanceTarget.sol

```diff
@@ -4,134 +4,61 @@ pragma solidity ^0.8.13;
-        require(msg.sender == address(connext), "Caller must be Connext");
-        _;
-    }
-
-    constructor(address _connext) {
-        connext = IConnext(_connext);
-    }
-
-    function xReceive(
-        bytes32 _transferId,
-        uint256 _amount, // Final Amount receive via Connext(After AMM calculation)
-        address _asset,
-        address _originSender,
-        uint32 _origin,
-        bytes calldata _callData
-    ) external returns (bytes memory) {
-        uint256 amount = _amount;
+    function _forwardFunctionCall(
+        bytes memory _preparedData,
+        bytes32,
+        uint256,
+        address
+    ) internal override returns (bool) {
         // Decode calldata
         (
+            uint256 amountOut,
+            address toAsset,
             uint24 poolFee,
             uint256 amountOutMin,
-            address from,
-            address to,
-            uint32 amountOfSwaps,
-            uint32 swapInterval,
-            address owner,
-            IDCAPermissionManager.PermissionSet[] memory permissions
-        ) = decode(_callData);
-
-        require(amount > 0, "!amount");
-        require(amountOutMin > 0, "!amountOut");
-        require(from != address(0), "!invalid");
-
-        if (from != _asset) {
-            // swap to deposit asset if needed
-            amount = swap(_asset, from, poolFee, amount, amountOutMin);
-            // TODO: add fallback to return funds to user address
-        }
-
-        // deposit
-        deposit(
-            from,
-            to,
-            amount,
-            amountOfSwaps,
-            swapInterval,
-            owner,
-            permissions
-        );
-    }
-
-    function encode(
-        uint24 poolFee,
-        uint256 amountOutMin,
-        address from,
-        address to,
-        uint32 amountOfSwaps,
-        uint32 swapInterval,
-        address owner,
-        IDCAPermissionManager.PermissionSet[] memory permissions
-    ) external pure returns (bytes memory) {
-        return
-            abi.encode(
-                poolFee,
-                amountOutMin,
-                from,
-                to,
-                amountOfSwaps,
-                swapInterval,
-                owner,
-                permissions
+            bytes memory forwardCallData
+        ) = abi.decode(
+                _preparedData,
+                (uint256, address, uint24, uint256, bytes)
             );
-    }
 
-    /// INTERNAL
-    function decode(
-        bytes calldata data
-    )
-        internal
-        pure
-        returns (
-            uint24 poolFee,
-            uint256 amountOutMin,
+        (
             address from,
             address to,
             uint32 amountOfSwaps,
             uint32 swapInterval,
             address owner,
             IDCAPermissionManager.PermissionSet[] memory permissions
-        )
-    {
-        (
-            poolFee,
-            amountOutMin,
+        ) = abi.decode(
```

Comment:
double decoding kind of annoying, is it strictly necessary? when decoding internally, is there another scheme to use besides `abi`?